mater/lib/src/lib.rs

```diff
@@ -18,16 +18,139 @@ mod unixfs;
 mod v1;
 mod v2;
 
+use std::{
+    collections::{HashMap, HashSet},
+    io::SeekFrom,
+};
+
 // We need to re-expose this because `read_block` returns `(Cid, Vec<u8>)`.
 pub use ipld_core::cid::Cid;
+use ipld_core::codec::Codec;
+use ipld_dagpb::DagPbCodec;
 pub use multicodec::{DAG_PB_CODE, IDENTITY_CODE, RAW_CODE};
-pub use stores::{create_filestore, Blockstore, Config, FileBlockstore};
+pub use stores::{
+    create_filestore, Blockstore, Config, FileBlockstore, DEFAULT_CHUNK_SIZE, DEFAULT_TREE_WIDTH,
+};
+use tokio::io::{AsyncReadExt, AsyncSeek, AsyncSeekExt, AsyncWriteExt};
 pub use v1::{Header as CarV1Header, Reader as CarV1Reader, Writer as CarV1Writer};
 pub use v2::{
     verify_cid, Characteristics, Header as CarV2Header, Index, IndexEntry, IndexSorted,
     MultihashIndexSorted, Reader as CarV2Reader, SingleWidthIndex, Writer as CarV2Writer,
 };
 
+/// Represents the location and size of a block in the CAR file.
+pub struct BlockLocation {
+    /// The byte offset in the CAR file where the block starts.
+    pub offset: u64,
+    /// The size (in bytes) of the block.
+    pub size: u64,
+}
+
+/// A simple blockstore backed by a CAR file and its index.
+pub struct CarBlockStore<R> {
+    reader: R,
+    /// Mapping from CID to block location.
+    pub index: HashMap<Cid, BlockLocation>,
+}
+
+impl<R> CarBlockStore<R>
+where
+    R: AsyncSeekExt + AsyncReadExt + Unpin,
+{
+    /// Extract content by traversing the UnixFS DAG using the index.
+    pub async fn extract_content_via_index<W>(
+        &mut self,
+        root: &Cid,
+        output: &mut W,
+    ) -> Result<(), Error>
+    where
+        W: AsyncWriteExt + Unpin,
+    {
+        // To avoid processing a block more than once.
+        let mut processed = HashSet::new();
+        // We use a stack for DFS traversal.
+        let mut to_process = vec![*root];
+
+        while let Some(current_cid) = to_process.pop() {
+            if processed.contains(&current_cid) {
+                continue;
+            }
+            processed.insert(current_cid);
+
+            // Retrieve block by CID via the index.
+            let block_bytes = self.get_block(&current_cid).await?;
+
+            // Write the raw block data. In a real UnixFS traversal you might need
+            // to reconstruct file content in order.
+            output.write_all(&block_bytes).await?;
```

Comment:
How is that not real UnixFS traversal?
I get confused, what extract_content_via_index is supposed to do?
What do we mean by index if actually we are not using index anywhere?

Reply:
its fine imo it was a leftover comment sry

Reply:
kk